src/ts_catalog/continuous_agg.c

```diff
@@ -1510,15 +1533,40 @@ generic_time_bucket_ng(const ContinuousAggsBucketFunction *bf, Datum timestamp)
 
 	if (strlen(bf->timezone) > 0)
 	{
-		return DirectFunctionCall3(ts_time_bucket_ng_timezone,
+		if (TIMESTAMP_NOT_FINITE(bf->origin))
+		{
+			/* using default origin */
+			return DirectFunctionCall3(ts_time_bucket_ng_timezone,
+									   IntervalPGetDatum(bf->bucket_width),
+									   timestamp,
+									   CStringGetTextDatum(bf->timezone));
+		}
+		else
+		{
+			/* custom origin specified */
+			return DirectFunctionCall4(ts_time_bucket_ng_timezone_origin,
+									   IntervalPGetDatum(bf->bucket_width),
+									   timestamp,
+									   TimestampGetDatum(bf->origin),
```

Comment:
@mfundul Why casting if we can just do it?

```suggestion
									   TimestampTzGetDatum(bf->origin),
```

At the end will be the same `Int64GetDatum` behind the scenes but make sense since we're using here `TIMESTAMP WITH TIMEZONE`

Reply:
Because this is already a hack, and we need to remind ourselves that:
1. it's not a `Timestamp`
2. the function does not take a `TIMESTAMP` datum argument
3. we are relying on an implementation detail

Reply:
I have to agree with @fabriziomello - there is no need in this casting here. The fact that both TimestampTz and Timestamp are stored as uint64_t UTC time in microseconds is very well-known, it's been this way for years and is very unlikely to change. What we actually have here is not two types, but two macros with different names that do exactly the same. There is no reason to prefer one or another. I just used Timestamp everywhere for consistancy.

Reply:
We can discuss about the cast, but there is also the `TimestampTzGetDatum()` function that should be used.

Reply:
Oh, OK. @fabriziomello do you suggest using `TimestampTzGetDatum` or `TimestampGetDatum` here? Personally, I'm fine with any option 🤷‍♂️

Reply:
Interestingly, in PostgreSQL there is TIMESTAMP_NOT_FINITE macro but there is no matching TZ-version of it.

Reply:
They still use `TimestampTz` where appropriate even if `TIMESTAMP_NOT_FINITE()` works for both: https://github.com/postgres/postgres/blob/REL_14_2/src/backend/utils/adt/timestamp.c#L5622

Reply:
After thinking a bit more about it I believe @mfundul is right. Here we should cast Timestamp to TimestampTz and use TimestampTzGetDatum(), because ts_time_bucket_ng_timezone_origin expects TimestampTz. I will update the pull request accordingly.